### ISSUE TITLE
Italicized text adding non-breaking spaces after publishing

### ISSUE THREAD

**Issue by edequalsawesome** [WordPress/gutenberg#13198]
**Describe the bug**
When writing a post in Gutenberg, if you italicize text, any spaces between the `<em>` tags are replaced with `&nbsp;`, which causes issues with formatting.

**To Reproduce**
1. Create a new post/page in Gutenberg
2. Italicize text using CTRL/⌘+I
3. Publish the post/page
4. Refresh the page to re-load the editor
5. View the block's HTML

**Expected behavior**
Italicized text would preserve spacing in `<em>` tags

**Actual behavior**
Spaces in text between the `<em>` tags are replaced with `&nbsp;`

**Screenshots**
![Creating a New Page](http://cld.wthms.co/Qulj2w+)

**Desktop (please complete the following information):**
 - OS: macOS (10.14.2)/Windows 10 (v.1809)
 - Browser: Safari, Chrome (macOS & Windows 10)
 - Version: Safari - 12.0.2, Chrome - 71.0.3578.98 (macOS & Windows 10)

**Additional context**
WordPress 5.0.2, hosted on Pressable

**Comment by azaozz** [comment]
This is not new and not only when using the `<em>` tag. The browsers have always inserted U+00A0 (nbsp) chars when the user types several spaces. At first only in contenteditable, then started to see that in textareas and even in text fields.

Lately Chrome/WebKit seem to be getting more aggressive with these nbsp chars. As far as I see they insert one as soon as the user presses the spacebar at the end of each word, and then replace it with a normal space U+0020 later on, when the user continues to type. However if the HTML tag is closed (and in some other seemingly random cases) that replacement doesn't seem to happen.

There is an old core trac ticket attempting to deal with this: https://core.trac.wordpress.org/ticket/26842. Unfortunately it was never implemented as it has the potential to break wpautop and introduce new hard to find/anticipate edge cases there. Perhaps we can revisit again considering Gutenberg doesn't (shouldn't) use wpautop any more.

An acceptable fix for the "overly-aggressive" U+00A0 invasion in Chrome would probably be to just replace them all with U+0020 on editor cleanup. That may eventually break somebody's "double-spacing-after-a-full-stop" style, but this would be the "lesser evil" :)

**Comment by iseulde** [comment]
This happens with any format. Looking into it a bit now.

**Comment by iseulde** [comment]
Interestingly, the `input` event is not triggered when you press the spacebar in a format.

**Comment by azaozz** [comment]
@iseulde As this is a "pure text" formatting, do you think something like this would be good?
```
content = content.replace( /(^|[^ \u00A0])\u00A0([^ \u00A0]|$)/g, ' ' );
```

It should replace single U+00A0 with U+0020 at the start, end or between any other chars.

**Comment by iseulde** [comment]
@azaozz There is still `contenteditable`, but I've made it so that the browser does not add any non breaking spaces when pressing space, by use of `white-space: pre-wrap;`.

**Comment by renemorozowich** [comment]
![em-bug](https://user-images.githubusercontent.com/26553939/51324994-02dd6580-1a3a-11e9-9052-4bf6e2a263f0.gif)
I'm having issues with this as well. I've tried using the keyboard shortcut for italics vs. using the italics button vs. putting in the em tags in the HTML manually.

**Comment by iseulde** [comment]
@azaozz If you add `white-space: pre-wrap;` to the `div`, the result when typing would look very different. It doesn't add any styling, it just changes how white space is rendered.

**Comment by aduth** [comment]
FYI there's the following note in the release note of TinyMCE 4.9.0 (newer than the 4.8.0 shipped currently in core):

>Reduced insertion of unnecessary nonbreaking spaces in the editor. #TINY-1879

(I had a hard time tracking this back to a specific code change; I guess the reference is some internal issue tracker?)

https://github.com/tinymce/tinymce/blob/master/changelog.txt

Manually pulling in the latest version, the issue still appears to be present though. 🤷‍♂️

**Comment by aduth** [comment]
> Manually pulling in the latest version, the issue still appears to be present though. 🤷‍♂️

Actually, on further testing, at least pressing Space at the end of an inline boundary in TinyMCE 4.9.2 does not insert the nbsp, whereas in 4.8.0 it would.

Related:

- https://github.com/tinymce/tinymce/commit/bad77bacb5113512c773be65d40c2b587cbd1c9d#diff-c51bc0bcc0d2d5008544b2cf3d001207
- https://github.com/tinymce/tinymce/commit/829942db6d935d5ff3b977f07c94b4c2fcbad979#diff-c51bc0bcc0d2d5008544b2cf3d001207
- https://github.com/tinymce/tinymce/commit/0167d682b9fc9b471a731c3cc8fcb5b620a91158#diff-c51bc0bcc0d2d5008544b2cf3d001207

**Comment by iseulde** [comment]
Ok, let's try updating TinyMCE in core then.

**Comment by azaozz** [comment]
> let's try updating TinyMCE in core then.

We could try, but... Don't think this will fix it, or at least not fully.

Here's an updated fiddle: https://jsfiddle.net/j5sLq4kf/. It clearly shows that Chrome inserts a nbsp as you type a space after any other char, then replaces that nbsp with a normal space when you continue to type.

Firefox always keeps a normal space at the end, but converts previous spaces into nbsp when typing several.

Also, the behaviour is a bit unpredictable when pasting in contenteditable.

Seems the best option is to add `white-space: pre-wrap;` and make sure no `\n` makes its way in there.

**Comment by AtrumGeost** [comment]
Hi folks!

I was able to replicate that using the `<strong>` tags:

![screen shot on 2019-01-28 at 12 48 42](https://user-images.githubusercontent.com/3696121/51855361-0eaf0e80-22fb-11e9-80ce-9478ae940413.png)

The steps I followed:

1. Created a draft with some text and used the "Bold" button. 
2. Closed the draft and then opened it again
3. Changed to the code editor

I used a Jurassic Ninja site to replicate that. And I wasn't able to replicate it when having the "Classic Editor" plugin enabled. Meaning this is only happening with Gutenberg
